app/src/main/java/net/measurementlab/ndt7/android/MainActivity.kt

```diff
@@ -44,6 +97,73 @@ class MainActivity : AppCompatActivity() {
 
         override fun onClientDownloadMeasurement(measurement: Measurement) {
             Log.d(TAG, "client measurement: $measurement")
+            Log.d(TAG,""+measurement.elapsed)
+
+            if(measurement.elapsed > 0) {
+                val rate: Double = (measurement.appInfo!!.numBytes).toDouble() / 2.0.pow(20.0) / (measurement.elapsed/ 1_000_000_000.0)
+
+                Thread(Runnable {
+                    this@MainActivity.runOnUiThread(java.lang.Runnable {
+                        val mb = measurement.appInfo!!.numBytes.toDouble() / 125000.0 / 10.0
+                        val upload = (10f - measurement.elapsed) * mb / (10f - measurement.elapsed)
+                        val uploadSpeedText = String.format(Locale.getDefault(), "%.2f MBit/s", upload)
+                        tv_uploadspeed.text = uploadSpeedText
+
+                    })
+                }).start()
+            }
+        }
+    }
+    private inner class MyClientKt constructor(settings: Settings) : Client(settings) {
```

Comment:
Why has the design now two different classes?

Reply:
Becuase I'm not familiar with Kotlin.

Reply:
IIRC you can use Kotlin classes from Java just fine. Isn't that the case?